src/components/PopupWithForm/PopupWithForm.js

```diff
@@ -11,36 +14,86 @@ export default function PopupWithForm({ name, onSwitch }) {
           onClick={() => onSwitch("")}
         />
         {name === "signup" && <FormSignUp onSwitch={onSwitch} />}
-        {name === "signin" && <FormSignIn onSwitch={onSwitch} />}
-        {name === "success" && <FormSucess onSwitch={onSwitch}/>}
+        {name === "signin" && <FormSignIn onSwitch={onSwitch} closeModal={() => onSwitch("")}/>}
+        {name === "success" && <FormSucess onSwitch={onSwitch} />}
       </div>
     </div>
   );
 }
 
 function FormSignUp({ onSwitch }) {
-  const [email, setEmail] = useState("");
-  const [password, setPassword] = useState("");
-  const [username, setUsername] = useState("");
+  const [tempUser, setTempUser] = useState({
+    email: "",
+    password: "",
+    username: "",
+  });
+  const [validation, setValidation] = useState({
+    emailValid: false,
+    passwordValid: false,
+    usernameValid: false,
+    formValid: false,
+  });
+  const [errors, setErrors] = useState({
+    email: "",
+    password: "",
+    username: "",
+  });
+
+  const validateField = (fieldName, value) => {
+    let fieldValidationErrors = errors;
+    let emailValid = validation.emailValid;
+    let passwordValid = validation.passwordValid;
+    let usernameValid = validation.usernameValid;
+
+    switch (fieldName) {
+      case "email":
+        emailValid = value.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);
+        fieldValidationErrors.email = emailValid ? "" : " is invalid";
+        break;
+      case "password":
+        passwordValid = value.length >= 8;
+        fieldValidationErrors.password = passwordValid ? "" : " is too short";
+        break;
+      case "username":
+        usernameValid = value.length >= 2;
+        fieldValidationErrors.username = usernameValid ? "" : " is too short";
+        break;
+      default:
+        break;
+    }
+
+    setValidation({
+      emailValid: emailValid,
+      passwordValid: passwordValid,
+      usernameValid: usernameValid,
+      formValid: emailValid && passwordValid && usernameValid,
+    });
 
-  const handleChangeEmail = (e) => {
-    setEmail(e.target.value);
+    setErrors(fieldValidationErrors);
   };
 
-  const handleChangePassword = (e) => {
-    setPassword(e.target.value);
+  const handleChange = (e) => {
+    const { name, value } = e.target;
+    validateField(name, value);
+    setTempUser({ ...tempUser, [name]: value });
   };
 
-  const handleChangeUsername = (e) => {
-    setUsername(e.target.value);
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    auth
+      .register(tempUser.email, tempUser.password, tempUser.username)
```

Comment:
**Might be better**

it's better to make api requests only in `App` because `Login` and `Register` should not know how to do it. They should be independent from such logics